src/js/components/conversation.jsx

```diff
@@ -54,34 +65,145 @@ export class ConversationComponent extends Component {
 
     };
 
+    onScroll = () => {
+        const {dispatch, shouldScrollToBottom} = this.props;
+
+        // If top of Conversation component is reached, we need to fetch older messages
+        const node = findDOMNode(this);
+        if(node.scrollTop === 0) {
+            this.fetchHistory();
+        } else if(shouldScrollToBottom) {
+            dispatch(setShouldScrollToBottom(false));
+        }
+    };
+
+    fetchHistory = () => {
+        const {dispatch, hasMoreMessages, isFetchingMoreMessages, messages} = this.props;
+
+        const node = findDOMNode(this);
+        if(hasMoreMessages && !isFetchingMoreMessages) {
+            // make sure the last message is one from the server, otherwise it doesn't need to scroll to previous first message
+            if (messages.length > 0 && messages[messages.length - 1]._id) {
+                this._lastTopMessageId = messages[0]._id;
+            }
+            
+            const top = getTop(this._topMessageNode, node);
+            this._lastTopMessageNodePosition = top - node.scrollTop;
+            dispatch(setFetchingMoreMessages(true));
+            setTimeout(() => {
+                fetchMoreMessages();
+            }, 400);
+        }
+    };
+
     scrollToBottom = () => {
-        const timeout = setTimeout(() => {
+        const {shouldScrollToBottom} = this.props;
+        if(!this._isScrolling && (shouldScrollToBottom || this._forceScrollToBottom)) {
+            this._isScrolling = true;
+            const timeout = setTimeout(() => {
```

Comment:
The reason why it's in a closure within setTimeout is because you want to have the `clientHeight` and `scrollHeight` when the content is flushed to the DOM. This was a valid thing to do back when we used Marionette and I assumed it still was necessary with React. Maybe it's not... we should probably test that out. 

But you are right that this could be made a class method if `setTimeout` is still a thing to do.